src/System.CommandLine/Parsing/StringExtensions.cs

```diff
@@ -322,32 +328,45 @@ void ReadResponseFile(string filePath, int i)
 
             if (args.Count > 0)
             {
-                try
+                if (inferRootCommand &&
+                    args[0] == RootCommand.ExecutablePath)
                 {
-                    potentialRootCommand = Path.GetFileName(args[0]);
+                    list.AddRange(args);
+                    return list;
                 }
-                catch (ArgumentException)
+                else
                 {
-                    // possible exception for illegal characters in path on .NET Framework
-                }
+                    try
+                    {
+                        potentialRootCommand = Path.GetFileName(args[0]);
 
-                if (potentialRootCommand != null &&
-                    commandLineConfiguration.RootCommand.HasAlias(potentialRootCommand))
-                {
-                    list.AddRange(args);
-                    return list;
+                        if (rootCommand.HasAlias(potentialRootCommand))
+                        {
+                            list.AddRange(args);
+                            return list;
+                        }
+                    }
+                    catch (ArgumentException)
+                    {
+                        // possible exception for illegal characters in path on .NET Framework
+                    }
                 }
             }
 
-            var commandName = commandLineConfiguration.RootCommand.Name;
+            var commandName = rootCommand.Name;
 
             list.Add(commandName);
 
             int startAt = 0;
 
-            if (FirstArgMatchesRootCommand())
+            var firstArgMatchesRootCommand = FirstArgMatchesRootCommand();
+
+            if (inferRootCommand)
             {
-                startAt = 1;
+                if (firstArgMatchesRootCommand)
```

Comment:
nit: Combine if statements to reduce nesting?
```C#
if (inferRootCommand && firstArgMatchesRootCommand)
```

Reply:
Or just delete the whole `if` block... :)